### PR TITLE
Create `call` method. Fixes #10

### LINKED ISSUE #10: gcx.call
A dev should be able to call a function from gcx.

Why? A HTTP trigger URL doesn't give errors whereas functions.call does.

```ts
const {call} = require('gcx');

async function main() {
  const {result, error} = await call({
    name: 'my-fn-name',
  });
  if (error) return console.error(error);
  console.log(result);
}
main();
```

See: https://cloud.google.com/functions/docs/reference/rest/v1/projects.locations.functions/call

### PR DESCRIPTION
**Pull request by grant**
Adds a `call` method for synchronously invoking a Cloud Function.
Fixes #10.

### CODE REVIEW

File: test/test.ts [hunk 1]
```diff
@@ -90,13 +89,19 @@ describe('cloud functions api', () => {
...
     const projectId = 'el-gato';
     const deployer = new gcx.Deployer({name, targetDir, projectId});
     await deployer.deploy();
     scopes.forEach(s => s.done());
   });
+
+  it('should call end to end', async () => {
+    mockUploadUrl(), mockUpload(), mockDeploy(), mockPoll();
+    const c = new gcx.Caller();
+    c.call(name);
```
Comment (JustinBeckwith) [inline test/test.ts]: It would be great to nock the calling endpoint, and assert that the response is what we expect.

File: src/index.ts [hunk 2]
```diff
@@ -265,23 +297,33 @@ export class Deployer extends EventEmitter {
...
-      google.options({auth});
-      this._gcf = google.cloudfunctions('v1');
-    }
-    return this._gcf;
+  async call(name: string) {
+    this.emit(ProgressEvent.STARTING);
+    const gcf = await this._getGCFClient();
+    const fns = gcf.projects.locations.functions;
+    this.emit(ProgressEvent.CALLING);
+    await fns.call({
```
Comment (JustinBeckwith) [inline src/index.ts]: Wouldn't we want to return the response from this?

File: src/index.ts [hunk 3]
```diff
@@ -265,23 +297,33 @@ export class Deployer extends EventEmitter {
...
 
 export async function deploy(options: DeployerOptions) {
   const deployer = new Deployer(options);
   return deployer.deploy();
 }
+
+export async function call(options: CallerOptions) {
+  const caller = new Caller();
+  return caller.call(options.functionName);
+}
```
Comment (JustinBeckwith) [inline src/index.ts]: please add a newline

**Comment by JustinBeckwith** [comment]
Sorry about the merge conflicts 😿 This was a problem on machines where the Cloud SDK was set up, and the user is logged in.  That's why it never popped up on CI. 

I submitted a fix just mocks out the auth library, skipping all that nock on GCE nonsense.  When you address the conflicts - let me know and we can merge!

**Comment by grant** [comment]
I'm still not sure how to fix the nock issues after merging with upstream.

I think I might just start with a fresh branch and retry a new PR unless you know the issue.

```sh
 6 passing (1s)
  3 failing

  1) cloud functions api
       should check to see if the function exists:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

false !== true

      + expected - actual

      -false
      +true

      at Context.it (test/test.ts:108:12)
      at process.internalTickCallback (internal/process/next_tick.js:77:7)

  2) end to end
       should deploy end to end:
     FetchError: request to https://oauth2.googleapis.com/token failed, reason: Nock: Disallowed net connect for "oauth2.googleapis.com:443/token"
      at ErroringClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1444:11)
      at ErroringClientRequest.<anonymous> (node_modules/nock/lib/intercept.js:232:10)
      at process.internalTickCallback (internal/process/next_tick.js:70:11)

  3) end to end
       should call end to end:
     FetchError: request to https://oauth2.googleapis.com/token failed, reason: Nock: Disallowed net connect for "oauth2.googleapis.com:443/token"
      at ErroringClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1444:11)
      at ErroringClientRequest.<anonymous> (node_modules/nock/lib/intercept.js:232:10)
      at process.internalTickCallback (internal/process/next_tick.js:70:11)
```

**Comment by JustinBeckwith** [comment]
Those errors make me thing you're not rebased against master *scratches head*.  Can you try switching to master, doing a pull, and running the tests with no changes?

**Comment by grant** [comment]
So I tried just creating a new branch:
- master: passes
- master + feat: passes
- master + feat + new test: fails

I'm not sure why there were 2 failing tests before, I must have not called a `super` or something.

```
  validation
    ✓ should throw if a name is not provided
    ✓ should throw if multiple triggers are provided

  ignore rules
    ✓ should return 0 rules if no .gcloudignore is availabe
    ✓ should return expected rules if .gcloudignore is availabe

  pack & upload
    ✓ should pack all of the files in the target dir (62ms)
    ✓ should PUT the file to Google Cloud Storage

  cloud functions api
    ✓ should check to see if the function exists

  end to end
    ✓ should work together end to end


  8 passing (133ms)
```

But if I add the new test it fails. Maybe there's something wrong with the test?

```ts
  it('should call end to end', async () => {
    const scopes =
        [mockUploadUrl(), mockUpload(), mockDeploy(), mockPoll(), mockCall()];
    const c = new gcx.Caller();
    const res = await c.call({functionName: name});
    assert.strictEqual(res.data.result, '{ "data": 42 }');
    scopes.forEach(s => s.done());
  });
```


```
1) end to end
       should call end to end:
     FetchError: request to https://oauth2.googleapis.com/token failed, reason: Nock: Disallowed net connect for "oauth2.googleapis.com:443/token"
      at ErroringClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1444:11)
      at ErroringClientRequest.<anonymous> (node_modules/nock/lib/intercept.js:232:10)
      at process.internalTickCallback (internal/process/next_tick.js:70:11)
```

**Comment by JustinBeckwith** [comment]
Ah yes!  This is super subtle, and probably bad on my part.  If your thingy requires auth, use `gcxp` instead of `gcx` to make the call.  The difference is that `gcxp` is proxyquired, overriding the auth library implementation.

**Comment by grant** [comment]
That's very tricky but makes sense.

I'm trying to nock the request by using the error in the nock but am failing find match the endpoint:

```ts
return nock('https://cloudfunctions.googleapis.com')
      .post('v1/%F0%9F%A6%84:call', {
        reqheaders: {
          "accept-encoding": [
            "gzip"
          ],
          "user-agent": [
            "google-api-nodejs-client/0.7.2 (gzip)"
          ],
          "accept": [
            "application/json"
          ],
          "content-length": [
            "0"
          ],
          "connection": [
            "close"
          ]
        }
      })
          // '/v1/projects/el-gato/locations/us-central1/functions/%F0%9F%A6%84:call')
      .reply(200, {
        executionId: 'my-execution-id',
        result: '{ "data": 42 }',
        error: null
      });
```


```
 FetchError: request to https://cloudfunctions.googleapis.com/v1/%F0%9F%A6%84:call failed, reason: Nock: No match for request {
  "method": "POST",
  "url": "https://cloudfunctions.googleapis.com/v1/%F0%9F%A6%84:call",
  "headers": {
    "accept-encoding": [
      "gzip"
    ],
    "user-agent": [
      "google-api-nodejs-client/0.7.2 (gzip)"
    ],
    "accept": [
      "application/json"
    ],
    "content-length": [
      "0"
    ],
    "connection": [
      "close"
    ]
  }
}
```

**Comment by JustinBeckwith** [comment]
I wouldn't bother with nocking anything beyond the url:

```js
return nock('https://cloudfunctions.googleapis.com').post('v1/*');
```

Then, when you get the response back I would check the `res.config` to ensure the url was what you expected it to be.

**Comment by grant** [comment]
Hmm, I still can't nock it.
The latest code is on `grant:grant_call2` if you can try.

**Comment by grant** [comment]
Closing in favor of #15.

**Pull request closed without merging**